### PR TITLE
[11.x] Implement condition helper

### PR DESCRIPTION
**Pull request by Anton5360**
### Description
Here is a small helper which might be handy when you need an if condition like this:
```php
    if ($user->isAdmin()) {
        return true;
    }

    return null;
```
So now you might do it in one line:
```php
    return condition($user->isAdmin());
```
### Benefits
Yes, I know about ternary operator existence :D, but:

1. IMO it just looks a bit clearer (I personally just don`t really like such ternary conditions)
```php
    return $user->isAdmin() ? null : false;
    -----
    return condition(value: $user->isAdmin(), truthy: null, falsy: false)
``` 
2. You might take advantage of passing a callable for any of parameters for additional logic:
```php
    condition(
        value: fn () => // Some difficult condition
        truthy: fn () => $this->resolveDifficultTruthyValue()
        falsy: fn () => throw new RuntimeException(),
    ) 
```
### Personal comments
This helper does not bring much functionality, it's basically the syntax sugar to improve coding experience (hopefully).
**P.S.** The most recent case when I would personally use it for my project is inside `before` method of Policy.

### Update
1. If only single argument is passed, then HigherOrderWhenProxy is returned
```php
    $test = new class {
        public function sayHi() 
        {
            return 'Hi!';
        }
    };
    
    condition($test)->sayHi();  // Hi!
    condition(fn () => $this->service->resolveTestOrNull())->sayHi();  // Returns "Hi!" if Test was resolved, null otherwise
```
2. Passes the resolved "value" parameter to both "truthy" and "falsy" (if they are callbacks)
```php
    condition(new stdClass(), truthy: fn ($stdClass) => ...);
    condition(fn () => new stdClass(), truthy: fn ($stdClass) => ...);
    condition(null, falsy: fn ($null) => ...);
```
3. Naming of the helper was changed to `condition` instead of `when`

### CODE REVIEW

**Comment by rodrigopedra** [comment]
Why default truthy value is `true`?

Why, in your first example, not just do:

```php
return $user->isAdmin() ?: null;
```

Also, the proposed implementation differs a lot from the `Illuminate\Support\Traits\Conditionable@when` method, which is already used in many places. That can be confusing.

**Comment by Anton5360** [comment]
> Why default truthy value is `true`?

I'm not sure actually about default values. I agree it might look kind of weird and confusing, so probably it would be better just to remove default values.
> Why, in your first example, not just do:
> 
> ```
> return $user->isAdmin() ?: null;
> ```

You are right, not the best example of mine. Let's change it a bit (same Policy `before` method):
```
    if ($user->isAdmin()) {
        return null;
    }

    return false;
    
    ------------
    
    return when(condition: $user->isAdmin(), truthy: null, falsy: false);
```
The obvious next question would be "why not to simply use ternary expression `$user->isAdmin() ? null : false;`", but again - it`s the simplest example just to demonstrate the idea. For example, you might also need to do something else before returning a value, so ternary operator would not work here.

> Also, the proposed implementation differs a lot from the `Illuminate\Support\Traits\Conditionable@when` method, which is already used in many places. That can be confusing.

I completely agree with you here. I didn't come up with a better naming yet, so if you or anyone else have any better suggestion, it will be much appreciated.
I was also wondering about such naming like "`bool_to()`", "`condition_to()`" or "`if_condition()`", something like that.



So basically, it may become handy if you want to make your code shorter (turn it into "one line") or to compact more complex conditions in one place (might be some callbacks with logic).

P.S. And thank you for your comments

**Comment by milwad-dev** [comment]
The `isAdmin` method return boolean and you don't need to if condition:
```php
return $user->isAdmin();
```

**Comment by crynobone** [comment]
Not a fan of this because it work differently from https://github.com/laravel/framework/blob/11.x/src/Illuminate/Conditionable/Traits/Conditionable.php, where else `tap()` and https://github.com/laravel/framework/blob/11.x/src/Illuminate/Support/Traits/Tappable.php does.

**Comment by donnysim** [comment]
To be fair, I will never understand the need for such helpers. For single lines it's useless and for multi lines for anything more serious it makes the indentation so much harder to discover where what starts or ends that maybe you just need to split it to smaller functions begin with. If it executes some logic you will always have to wrap it with a closure and pass all the outer variables etc. and if it fits into a single line, why would you prefer this over a ternary. I know laravel community is crazy about such utilities and would not be surprised if this is accepted.

**Comment by Anton5360** [comment]
> The `isAdmin` method return boolean and you don't need to if condition:
> 
> ```
> return $user->isAdmin();
> ```

Please, consider checking a description firstly :)

**Comment by Anton5360** [comment]
> Not a fan of this because it work differently from https://github.com/laravel/framework/blob/11.x/src/Illuminate/Conditionable/Traits/Conditionable.php, where else `tap()` and https://github.com/laravel/framework/blob/11.x/src/Illuminate/Support/Traits/Tappable.php does.

I see your point. I mean, it should be better if we come up with a better naming for it. I'm not proud of `when` naming as it poorly describes what this small function does let alone making it confusing with built-in Tappable and Conditionable Traits.

It might be something more straight forward, I left a couple of my suggestions in the comment above, but still might be anything better

**Comment by Anton5360** [comment]
> To be fair, I will never understand the need for such helpers. For single lines it's useless and for multi lines for anything more serious it makes the indentation so much harder to discover where what starts or ends that maybe you just need to split it to smaller functions begin with. If it executes some logic you will always have to wrap it with a closure and pass all the outer variables etc. and if it fits into a single line, why would you prefer this over a ternary. I know laravel community is crazy about such utilities and would not be surprised if this is accepted.

If for `except` and `only` methods for Enum rule I was like "okay, this can be super useful", so when I came up with this PR I was like "okay, someone might find it useful for some case. Let me suggest this and take a look what others think".
I'm not a big fan of helpers either, I personally almost never use them, but I personally would likely find use cases for this one.

Not saying this is the most must have feature which I`m gonna use all day every day. It`s just a helper which might help you sometimes, that`s it.

If no one sees the case when this thingy may become handy, we can always just close the PR )

**Comment by Anton5360** [comment]
> and if it fits into a single line, why would you prefer this over a ternary

For example, when I have a ternary expression like
```
    return $this->someClassName->longMethodNameAcceptingManyParams(
        param1: $param1,
        param2: $param2,
        param3: $param3
    ) 
        ? $collection->some()
            ->chain()
            ->with()
            ->logic()
        : $this->someOtherClass->doSomethingElse(
            param1: $param1,
            param2: $param2,
            param3: $param3    
        );
```
It's a made up example but you see the point. Personally, I would never go with such ternary expression, IMPO it's super unreadable. However, with "syntax sugar" helper `when()` (naming is a nightmare, please feel free to suggest something better:)) we might do it more readable:
```
    return when(
        condition: $this->someClassName->longMethodNameAcceptingManyParams(
            param1: $param1,
            param2: $param2,
            param3: $param3
        ), 
        truthy: static fn () => $collection->some()
                ->chain()
                ->with()
                ->logic(),
        falsy: fn () => $this->someOtherClass->doSomethingElse(
                    param1: $param1,
                    param2: $param2,
                    param3: $param3    
                )
    );
```

That`s basically the main purpose of this small helper

**Comment by donnysim** [comment]
I get your point, but even these examples don't look great either with ternary nor with the helper to me, simple if statement would be the way to go, sometimes even with a variable that makes it easier to debug later if needed, and sometimes you need the result of that said condition inside your if branches that slims the use case even more - unless you pass the result of the condition to the callbacks.

**Comment by rodrigopedra** [comment]
I tried to mimic `Illuminate\Support\Traits\Conditionable@when` as possible, to propose an alternative:

```php
<?php // ./routes/console.php

use Illuminate\Support\Facades\Artisan;
use Illuminate\Support\HigherOrderWhenProxy;

function when($value, $truthy = null, $falsy = null)
{
    $value = value($value);

    if (func_num_args() === 1) {
        return (new HigherOrderWhenProxy($value))->condition($value);
    }

    return $value
        ? value($truthy, $value)
        : value($falsy ?? $value, $value);
}

class Foo
{
    public function __construct(
        public readonly string $foo,
    ) {}

    public function bar(): string
    {
        return $this->foo . 'bar';
    }
}

Artisan::command('app:test', function () {
    $this->info('Single argument -> proxy');
    dump(
        when(new Foo('foo'))->bar(),
        when(null)->bar(),
    );

    $this->info('Two argument -> default falsy');
    dump(
        when(null, true),
        when(false, true),
        when(new Foo('foo'), 'Yay!'),
        when(true, fn () => 'lazy'),
    );

    $this->info('Three arguments');
    dump(
        when(null, true, false),
        when(new Foo('foo'), true, false),
        when(
            new Foo('foo'),
            fn ($foo) => $foo->bar(),
            fn () => throw new \RuntimeException('lazy'),
        ),
    );
});
```

Results

```shell
$ php artisan app:test
Single argument -> proxy
"foobar"
null
Two argument -> default falsy
null
false
"Yay!"
"lazy"
Three arguments
false
true
"foobar"
```

**Comment by Anton5360** [comment]
@rodrigopedra It's definitely better than the my initial implementation, good suggestion imo.
The only thing I don't agree with in your implementation is this `$falsy ?? $value`. I believe `$falsy` should be returned always

**Comment by Anton5360** [comment]
We`ve got mostly negative feedback regarding this PR, so I`m quite lost if this thing still has a chance to be merged.
 
Does it still make sense to implement your suggestion @rodrigopedra, what do you think?

**Comment by Anton5360** [comment]
@rodrigopedra good suggestion, I`ve implemented it and updated description

**Comment by Anton5360** [comment]
> I get your point, but even these examples don't look great either with ternary nor with the helper to me, simple if statement would be the way to go, sometimes even with a variable that makes it easier to debug later if needed, and sometimes you need the result of that said condition inside your if branches that slims the use case even more - unless you pass the result of the condition to the callbacks.

Now it got some improvement (thanks to @rodrigopedra's suggestion). What do you think now, does it seem more useful? 
I`m interested in your opinion as maybe you might suggest as well how to make it worthy to be merged into framework ecosystem

**Comment by crynobone** [comment]
Hi @Anton5360, can you rebase this PR to latest upstream

**Comment by taylorotwell** [comment]
Thanks for your pull request to Laravel!

Unfortunately, I'm going to delay merging this code for now. To preserve our ability to adequately maintain the framework, we need to be very careful regarding the amount of code we include. 

If applicable, please consider releasing your code as a package so that the community can still take advantage of your contributions!
